src/util/timer.rs

```diff
@@ -0,0 +1,254 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::mem::replace;
+use std::cmp::Ordering;
+use std::time::Duration;
+use std::collections::BinaryHeap;
+
+use util::time::Instant;
+
+#[derive(Debug)]
+pub struct TimeoutTask<T> {
+    next_tick: Instant,
+    timeout: Duration,
+    task: T,
+}
+
+impl<T> TimeoutTask<T> {
+    pub fn into_inner(self) -> T {
+        self.task
+    }
+    pub fn timeout(&self) -> Duration {
+        self.timeout
+    }
+}
+
+impl<T> AsRef<T> for TimeoutTask<T> {
+    fn as_ref(&self) -> &T {
+        &self.task
+    }
+}
+
+impl<T> AsMut<T> for TimeoutTask<T> {
+    fn as_mut(&mut self) -> &mut T {
+        &mut self.task
+    }
+}
+
+impl<T> PartialEq for TimeoutTask<T> {
+    fn eq(&self, other: &TimeoutTask<T>) -> bool {
+        if self.next_tick == other.next_tick {
+            return true;
+        }
+        false
+    }
+}
+
+impl<T> Eq for TimeoutTask<T> {}
+
+impl<T> PartialOrd for TimeoutTask<T> {
+    fn partial_cmp(&self, other: &TimeoutTask<T>) -> Option<Ordering> {
+        Some(if self.next_tick > other.next_tick {
+            Ordering::Less
+        } else if self.next_tick < other.next_tick {
+            Ordering::Greater
+        } else {
+            Ordering::Equal
+        })
+    }
+}
+
+impl<T> Ord for TimeoutTask<T> {
+    fn cmp(&self, other: &TimeoutTask<T>) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+pub struct Timer<T> {
+    pending: BinaryHeap<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    pub fn new(capacity: usize) -> Self {
+        Timer {
+            pending: BinaryHeap::with_capacity(capacity),
+        }
+    }
+
+    /// Add a periodic task into the `Timer`. The `timeout` will
+    /// be aligned up to `tick` of the `Timer` if need.
+    pub fn add_task(&mut self, timeout: Duration, task: T) {
+        let task = TimeoutTask {
+            next_tick: Instant::now() + timeout,
+            timeout: timeout,
+            task: task,
+        };
+        self.pending.push(task);
+    }
+
+    /// Remove a task from the timer. Returns the `TimeoutTask` if found.
+    pub fn remove_task<F>(&mut self, f: F) -> Option<TimeoutTask<T>>
+    where
+        F: Fn(&TimeoutTask<T>) -> bool,
+    {
+        let mut res = None;
+        let mut vec = replace(&mut self.pending, BinaryHeap::new()).into_vec();
+        if let Some((idx, _)) = vec.iter().enumerate().find(|&(_, task)| f(task)) {
+            res = Some(vec.swap_remove(idx));
+        }
+        self.pending = BinaryHeap::from(vec);
+        res
+    }
+
+    /// Get the next `timeout` from the timer.
+    pub fn next_timeout(&mut self) -> Option<Instant> {
+        self.pending.peek().map(|task| task.next_tick)
+    }
+
+    pub fn tasks_before(&mut self, instant: Instant) -> TimerTaskBefore<T> {
```

Comment:
do we really need this? a helper function to get all timeout tasks is enough?

Reply:
Suppose this use case:
```rust
    let timeout  = timer.next_timeout();
    sleep(now() - timeout);
    let tasks: Vec<_> = timer.tasks_before(timeout);  // `tasks_before` is a helper function
    for (i, task) in  tasks {
        if tasks.should_remove_next_task() {
            timer.remove_task(tasks[i + 1]);    // It can't remove the target task from `tasks`.
        }
    }
```

Reply:
Em, so why not 

```
while let (task) = timer.try_pop(now) {
     on_timer(timer, task)
}
```